vllm/model_executor/parameter.py

```diff
@@ -16,329 +17,378 @@
-        param_data = param_data.narrow(self.output_dim, shard_offset,
+        param_data = param_data.narrow(param.output_dim, shard_offset,
                                        shard_size)
-        loaded_weight = loaded_weight.narrow(self.output_dim,
+        loaded_weight = loaded_weight.narrow(param.output_dim,
                                              shard_id * shard_size, shard_size)
 
         assert param_data.shape == loaded_weight.shape
         param_data.copy_(loaded_weight)
 
+    param.output_dim = output_dim
+    param.load_column_parallel_weight = lambda loaded_weight: (
+        load_column_parallel_weight(param, loaded_weight))
+    param.load_merged_column_weight = lambda loaded_weight, **kwargs: (
+        load_merged_column_weight(param, loaded_weight, **kwargs))
+    param.load_qkv_weight = lambda loaded_weight, **kwargs: (load_qkv_weight(
+        param, loaded_weight, **kwargs))
+    add_param_feature(param, Features.Column)
+
+
+def RowvLLMParameter(data: torch.Tensor, **kwargs) -> Parameter:
+    param = Parameter(data, requires_grad=False)
+    wrap_base_vllm_parameter(param, **kwargs)
+    wrap_row_vllm_parameter(param, **kwargs)
+    return param
+
 
-class RowvLLMParameter(BasevLLMParameter):
+def wrap_row_vllm_parameter(param: Parameter, input_dim: int,
+                            **kwargs) -> None:
     """
-    Parameter class defining weight_loading functionality
-    (load_row_parallel_weight) for parameters being loaded
-    into linear layers with row parallel functionality.
-    Requires an input_dim to be defined.
+    Add functionality to the parameter for loading weights into
+    linear layers with row parallelism. This includes layers
+    which are fused on disk. Requires an input dimension to be
+    defined. Called within the weight loader of each of the
+    row parallel linear layers.
     """
 
-    def __init__(self, input_dim: int, **kwargs):
-        self._input_dim = input_dim
-        super().__init__(**kwargs)
+    def load_row_parallel_weight(param: Parameter,
+                                 loaded_weight: torch.Tensor):
+        tp_rank = get_tensor_model_parallel_rank()
+        shard_size = param.data.shape[input_dim]
+        loaded_weight = loaded_weight.narrow(input_dim, tp_rank * shard_size,
+                                             shard_size)
+        assert param.data.shape == loaded_weight.shape
+        param.data.copy_(loaded_weight)
+
+    param.input_dim = input_dim
+    param.load_row_parallel_weight = lambda loaded_weight: (
+        load_row_parallel_weight(param, loaded_weight))
+    add_param_feature(param, Features.Row)
+
+
+def ModelWeightParameter(data: torch.Tensor, **kwargs) -> Parameter:
+    param = Parameter(data, requires_grad=False)
+    wrap_base_vllm_parameter(param, **kwargs)
+    wrap_column_vllm_parameter(param, **kwargs)
+    wrap_row_vllm_parameter(param, **kwargs)
+    wrap_model_weight_parameter(param, **kwargs)
```

Comment:
since we're not longer inheriting, I don't think we actually need `wrap_model_weight_parameter` just to add the param feature. 

We can keep for now likely not necessary